### ISSUE TITLE
Is there a way to support Non-greedy match?

### ISSUE THREAD

**Issue by 745198699** [skvadrik/re2c#208]
For example "a+?"

**Comment by skvadrik** [comment]
Short answer: no, re2c currently has no support for non-greedy match.

Long answer: this is probably not what you need, but non-greedy *submatch* extraction can be done with re2c `--tags` option by putting non-consuming part of alternative before the consuming part: `""|"a"`, as in this example:
```
#include <stdio.h>

static int lex(const char *YYCURSOR)
{
    const char *t1, *t2, *t3;
    /*!stags:re2c format = 'const char *@@;'; */
    /*!re2c
        re2c:define:YYCTYPE = char;
        re2c:yyfill:enable = 0;

        nongreedy = (""|"a")*;

        @t1 nongreedy @t2 nongreedy @t3 {
            printf("1st: '%.*s', 2nd: '%.*s'\n", t2 - t1, t1, t3 - t2, t2);
            return 0;
        }
    */
}

int main(int argc, char **argv)
{
    for (int i = 1; i < argc; ++i) lex(argv[i]);
    return 0;
}
```
If I run it on a string `"aaaaa"`, all `'a'` characters are matched by the second part of the rule (after tag `@t1`:
```
$ re2c test.re -T -o test.c && gcc test.c -otest && ./test aaaaa
1st: '', 2nd: 'aaaaa'
```
However, there's no way to make the *whole match* non-greedy. The lexer will still consume all `'a'`-s, you can only control the way it parses the consumed string (where it decides to put tags).

**Comment by olleharstedt** [comment]
This is still true?

**Comment by pmetzger** [comment]
Generally, I've found that non-greedy stuff I've wanted has almost always been possible to express with greedy operators and some care. It's not the worst limitation. It's true that it's nicer (for example) to express things like string syntax with non-greedy operators, but you don't strictly need it.

**Comment by olleharstedt** [comment]
Multiline comments were the use-case I was thinking of. :)

**Comment by skvadrik** [comment]
> Multiline comments were the use-case I was thinking of. :)

You can accurately lex multiline comments (nested or not) if you split the automaton into multiple automata (either using multiple re2c blocks, or conditions). The first block (start condition) would lex comment start and the second one would loop lexing comment body, or stop when it encounters comment end. This way you can handle escape sequences, end of input in a comment, and so on.

Here is a simple example with blocks:
```cpp
#include <assert.h>

bool is_comment(const char* YYCURSOR) {
/*!re2c
    re2c:yyfill:enable = 0;
    re2c:define:YYCTYPE = char;

    [/][*] { goto comment; }
    *      { return false; }
*/
comment:
/*!re2c
    [\x00] { return false; }
    [*][/] { return true; }
    [\][*] { goto comment; }
    *      { goto comment; }
*/
}

int main() {
    assert(is_comment("/**/"));
    assert(is_comment("/*aaa*/"));
    assert(is_comment("/*aaa\\*/aa*/"));
    assert(!is_comment(""));
    assert(!is_comment("/*/"));
    assert(!is_comment("/*aaa"));
    return 0;
}
```

And here's one with conditions:
```cpp
#include <assert.h>

/*!conditions:re2c*/
#define YYGETCONDITION() cond
#define YYCTYPE          char

bool is_comment(const char* YYCURSOR) {
    int cond = yycstart;
/*!re2c
    re2c:yyfill:enable = 0;

    <start> [/][*] { goto yyc_comment; }
    <start> *      { return false; }

    <comment> [\x00] { return false; }
    <comment> [*][/] { return true; }
    <comment> [\][*] { goto yyc_comment; }
    <comment> *      { goto yyc_comment; }
*/
}

int main() {
    assert(is_comment("/**/"));
    assert(is_comment("/*aaa*/"));
    assert(is_comment("/*aaa\\*/aa*/"));
    assert(!is_comment(""));
    assert(!is_comment("/*/"));
    assert(!is_comment("/*aaa"));
    return 0;
}
```

It may look a bit bulky, but in real life if it's part of a larger lexer that has to handle complex lexemes like strings with escape sequences, it is usually ok to have subautomata for parts of grammar that cannot be handled with a single regexp.

**Comment by pmetzger** [comment]
I'm a crazy person; I have built single regexps for multi-line comments and used them successfully. The regexp for C-style multi-line comments isn't too bad and is kind of a fun exercise.

I've also built them for strings complete with escape sequences, but I then abandoned them because it turned out that I needed to process the escape sequences anyway, and doing it in a subautomaton was easier and also allowed me to share code between multiple kinds of strings.
